src/SAML2/XML/md/AbstractMetadataDocument.php

```diff
@@ -0,0 +1,212 @@
+<?php
+
+namespace SAML2\XML\md;
+
+use DOMElement;
+use SAML2\SignedElementTrait;
+use SAML2\Utils;
+use SAML2\XML\ExtendableElementTrait;
+
+/**
+ * Class to represent a metadata document
+ *
+ * @package simplesamlphp/saml2
+ */
+abstract class AbstractMetadataDocument extends AbstractSignedMdElement
+{
+    use ExtendableElementTrait;
+    use SignedElementTrait;
+
+    /**
+     * The ID of this element.
+     *
+     * @var string|null
+     */
+    protected $ID;
+
+    /**
+     * How long this element is valid, as a unix timestamp.
+     *
+     * @var int|null
+     */
+    protected $validUntil;
+
+    /**
+     * The length of time this element can be cached, as string.
+     *
+     * @var string|null
+     */
+    protected $cacheDuration;
+
+
+    /**
+     * Generic constructor for SAML metadata documents.
+     *
+     * @param string|null $ID The ID for this document. Defaults to null.
+     * @param int|null    $validUntil Unix time of validity for this document. Defaults to null.
+     * @param string|null $cacheDuration Maximum time this document can be cached. Defaults to null.
+     * @param \SAML2\XML\md\Extensions[]|null $extensions An array of extensions. Defaults to null.
+     */
+    public function __construct(
+        ?string $ID = null,
+        ?int $validUntil = null,
+        ?string $cacheDuration = null,
+        ?array $extensions = null
+    ) {
+        $this->ID = $ID;
+        $this->setValidUntil($validUntil);
+        $this->cacheDuration = $cacheDuration;
+        $this->setExtensions($extensions);
+    }
+
+
+    /**
+     * Process an XML element and get its ID property, if any.
+     *
+     * @param \DOMElement $xml An element that may contain an ID.
+     *
+     * @return string|null
+     */
+    public static function getIDFromXML(DOMElement $xml): ?string
```

Comment:
We should get rid of this one and use `self::getAttribute()` instead.

Reply:
Removed